### PR TITLE
Routing topic updates

### LINKED ISSUE #8763: Update routing docs
*From @JamesNK (https://github.com/aspnet/Routing/issues/720)* ...

- [x] **catchAll @JamesNK (https://github.com/aspnet/Docs/pull/8759)
- [x] Discuss MVC's EnableEndpointRouting flag @guardrex
- [x] Conceptual differences between Endpoints and Routes @rynowak
- [x] Differences in behaviour between Endpoints and Routes (mostly link generation) @rynowak
- [x] Link Generator @rynowak
- [x] Extensibility via Parameter Transformers @JamesNK (https://github.com/aspnet/Docs/pull/8759)
- [x] Extensibility via Endpoint Selector @rynowak

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 32374077-e02e-d6a6-1cea-05e58447b7cd
* Version Independent ID: 86aeab13-1c25-9fc1-4e56-0390cb40c242
* Content: [Routing in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-2.1)
* Content Source: [aspnetcore/fundamentals/routing.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/fundamentals/routing.md)
* Product: **aspnet-core**
* GitHub Login: @ardalis
* Microsoft Alias: **riande**

### LINKED ISSUE #9570: Unclear what routes variable means
There are many examples that use routes.MapRoute without mentioning what the routes variable mean.  One can only guess from the Use Routing Middleware sample in the middle of the article that it is probably a RouteBuilder instance.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 32374077-e02e-d6a6-1cea-05e58447b7cd
* Version Independent ID: 86aeab13-1c25-9fc1-4e56-0390cb40c242
* Content: [Routing in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-2.1)
* Content Source: [aspnetcore/fundamentals/routing.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/fundamentals/routing.md)
* Product: **aspnet-core**
* GitHub Login: @ardalis
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #8763 
Fixes #9570

* Versioning this for the 2.2 topic to only explain the endpoint routing scenario. The bit at the top on `EnableEndpointRouting` says that to learn more about legacy routing to see the 2.1 version of the topic.
* URI/URL ... I didn't make radical changes; but AFAICT, this needs work. Should I convert all instances of "URL" to the more general "URI" then you can flag the ones that go back to "URL"? Otherwise, let me know which "URL" mentions are better as URI.
* No doubt some of the versioning and member descriptions are incorrect in spots. I couldn't tell in some cases if/how the member works with endpoint routing. *There be 🐉 dragons 🐉 in my guesses!*
* Normally, we try not to refer to specific versions; however, this is a case where it can't be easily avoided: The endpoint routing paradigm prompts enough shifts in dev thinking that we probably need to point out the deltas explicitly. In some future version ... ASP.NET Core 3.0 is a logical time ... I think these callouts for 'this is new in 2.2' can be removed. I created #9552 to track the work for 3.0.
* IMO, the docs are 🎷 *jam'in 2.2 updates* 🎸, so we're not in a super-major-crazy hurry for the reviews.
* Let's avoid the 🎄 holiday rush 🎅 and take the sample to 2.2 now. I'll mark it on the sample update tracking issue shortly.

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@JamesNK is this something you could review if @rynowak 	is too busy?

File: aspnetcore/fundamentals/routing.md [hunk 1]
```diff
@@ -153,168 +430,147 @@ routes.MapRoute(
...
+URL patterns that attempt to capture a file name with an optional file extension have additional considerations. For example, consider the template `files/{filename}.{ext?}`. When values for both `filename` and `ext` exist, both values are populated. If only a value for `filename` exists in the URL, the route matches because the trailing period (`.`) is  optional. The following URLs match this route:
 
 * `/files/myFile.txt`
 * `/files/myFile`
 
-You can use the `*` character as a prefix to a route parameter to bind to the rest of the URI. This is called a *catch-all* parameter. For example, `blog/{*slug}` matches any URI that starts with `/blog` and had any value following it (which is assigned to the `slug` route value). Catch-all parameters can also match the empty string.
-
 ::: moniker range=">= aspnetcore-2.2"
 
+You can use an asterisk (`*`) or double asterisk (`**`) as a prefix to a route parameter to bind to the rest of the URI. These are called a *catch-all* parameters. For example, `blog/{**slug}` matches any URI that starts with `/blog` and has any value following it, which is assigned to the `slug` route value. Catch-all parameters can also match the empty string.
```
Comment (JamesNK) [inline aspnetcore/fundamentals/routing.md]: Maybe link to somewhere that discusses the difference between one and two asterisks so devs aren't confused why there are two syntaxes for the same thing

Comment (JamesNK) [reply aspnetcore/fundamentals/routing.md]: It is literally in the next paragraph. Never mind! 🤣

File: aspnetcore/fundamentals/routing.md [hunk 2]
```diff
@@ -153,168 +430,147 @@ routes.MapRoute(
...
+Route parameters may have constraints that must match the route value bound from the URL. Adding a colon (`:`) and constraint name after the route parameter name specifies an *inline constraint* on a route parameter. If the constraint requires arguments, they're enclosed in parentheses (`(...)`) after the constraint name. Multiple inline constraints can be specified by appending another colon (`:`) and constraint name.
```
Comment (JamesNK) [inline aspnetcore/fundamentals/routing.md]: Add a quick mention that constraints are used in link generation as well

File: aspnetcore/fundamentals/routing.md [hunk 3]
```diff
@@ -153,168 +430,147 @@ routes.MapRoute(
...
 
 ::: moniker range=">= aspnetcore-2.2"
 
-Route parameters may also have parameter transformers, which transform a parameter's value when generating links and matching actions and pages to URIs. Like constraints, parameter transformers can be added inline to a route parameter by adding a colon (`:`) and transformer name after the route parameter name. For example, the route template `blog/{article:slugify}` specifies a `slugify` transformer.
+Route parameters may also have parameter transformers, which transform a parameter's value when generating links and matching actions and pages to URLs. Like constraints, parameter transformers can be added inline to a route parameter by adding a colon (`:`) and transformer name after the route parameter name. For example, the route template `blog/{article:slugify}` specifies a `slugify` transformer.
```
Comment (JamesNK) [inline aspnetcore/fundamentals/routing.md]: Add a link to the section with more details on transformers.

File: aspnetcore/fundamentals/routing.md [hunk 4]
```diff
@@ -328,18 +584,18 @@ The following keywords are reserved names and can't be used as route names or pa
 
 ## Route constraint reference
 
-Route constraints execute when a `Route` has matched the syntax of the incoming URL and tokenized the URL path into route values. Route constraints generally inspect the route value associated via the route template and make a yes/no decision about whether or not the value is acceptable. Some route constraints use data outside the route value to consider whether the request can be routed. For example, the <xref:Microsoft.AspNetCore.Routing.Constraints.HttpMethodRouteConstraint> can accept or reject a request based on its HTTP verb.
+Route constraints execute when a match has occurred to the incoming URL and the URL path is tokenized into route values. Route constraints generally inspect the route value associated via the route template and make a yes/no decision about whether or not the value is acceptable. Some route constraints use data outside the route value to consider whether the request can be routed. For example, the <xref:Microsoft.AspNetCore.Routing.Constraints.HttpMethodRouteConstraint> can accept or reject a request based on its HTTP verb.
```
Comment (JamesNK) [inline aspnetcore/fundamentals/routing.md]: Mention that route constraints are used in link generation as well.

File: aspnetcore/fundamentals/routing.md [hunk 5]
```diff
@@ -1,77 +1,319 @@
...
+    }
+
+    public async Task InvokeAsync(HttpContext httpContext)
+    {
+        var url = _linkGenerator.GenerateLink(new { controller = "Store",
+                                                    action = "ListProducts" });
+
+        httpContext.Response.ContentType = "text/plain";
+
+        return httpContext.Response.WriteAsync($"Go to {url} to see our products.");
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply aspnetcore/fundamentals/routing.md]: (or leave the return and remove the async)

File: aspnetcore/fundamentals/routing.md [hunk 6]
```diff
@@ -1,77 +1,319 @@
...
 
 > [!IMPORTANT]
-> This document covers low-level ASP.NET Core routing. For information on ASP.NET Core MVC routing, see <xref:mvc/controllers/routing>.
+> This document covers low-level ASP.NET Core routing. For information on ASP.NET Core MVC routing, see <xref:mvc/controllers/routing>. For information on routing conventions in Razor Pages, see <xref:razor-pages/razor-pages-conventions>.
 
 [View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/fundamentals/routing/samples) ([how to download](xref:index#how-to-download-a-sample))
 
 ## Routing basics
 
+Most apps should choose a basic and descriptive routing scheme so that URLs are readable and meaningful. The default conventional route for ASP.NET Core MVC apps (`{controller=Home}/{action=Index}/{id?}`):
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: I think this is just here to describe the default conventional route. It merely says that it's a starting point.

Attribute routing is mentioned and linked in the next sentence. However, I don't like the way that the subject makes it sound like attribute routing is only for Web APIs, so I'll change that and then let's take a look on the next commit.

File: aspnetcore/fundamentals/routing.md [hunk 7]
```diff
@@ -1,77 +1,319 @@
...
 
 ### URL matching
 
+::: moniker range=">= aspnetcore-2.2"
+
+URL matching is the process by which routing dispatches an incoming request to an *endpoint*. This process is based on data in the URL path but can be extended to consider any data in the request. The ability to dispatch requests to separate handlers is key to scaling the size and complexity of an app.
+
+When a Routing Middleware executes, it sets an endpoint (`Endpoint`) and route values to a feature on the <xref:Microsoft.AspNetCore.Http.HttpContext>. Middleware running after the Routing Middleware can see the endpoint and can take actions based on the middleware. For example, an authorization middleware can interrogate the endpoint's metadata collection for an authorization policy. After all of the middleware in the request processing pipeline is executed, the selected endpoint's delegate is invoked.
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: https://github.com/aspnet/Docs/issues/9558

File: aspnetcore/fundamentals/routing.md [hunk 8]
```diff
@@ -1,77 +1,319 @@
...
+
+  ```cshtml
+  @page "{id?}"
+  ```
+
+  If the URI is `/Store/Product/18` in ASP.NET Core 2.1 or earlier, the link generated on the Store/Info page by `@Url.Page("/Login")` is `/Login/18`. The `id` value of 18 is reused, even though the link destination is different part of the app entirely. The `id` route value in the context of the `/Login` page is probably a user ID value, not a store product ID value.
+
+  In endpoint routing with ASP.NET Core 2.2 or later, the result is `/Login`. Ambient values aren't reused when the linked destination is a different action or page.
+
+* Round-tripping route parameter syntax: Forward slashes aren't encoded when using a double-asterisk (`**`) catch-all parameter syntax.
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: I'll add a remark and a note on the issue I opened earlier that the remark should be pulled later. https://github.com/aspnet/Docs/issues/9552

File: aspnetcore/fundamentals/routing.md (hunk 2 above)
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: I'll add it to the first paragraph of the *Route constraint reference* section. This bit (like an earlier section in the topic) makes some constraint comments and then links down to that section. Let's take a look on the next commit and see if it works.

**Comment by guardrex** [comment]
Thanks @JamesNK 

> Nice! Nothing major

Not so many 🐉 after all is good news. Let's tell management that I'm getting better ... well ... we do live in the age of *alternative facts* 😝 ... so we should be able to get away with it. 😨

It probably needs a bit more work on a couple of those touch-ups. For example, the bit in the *Routing basics* section may need more work.

**Comment by guardrex** [comment]
I'll be adding something this morning (in a few hours) to address feedback on #9570. I'll WIP this until then.

**Comment by guardrex** [comment]
I'll need an engineering sign-off in order to proceed. No giant rush tho ... AFAICT, we're look'in real good on 2.2 work.

File: aspnetcore/fundamentals/routing.md [hunk 9]
```diff
@@ -1,77 +1,326 @@
...
+```
+
+The `EnableEndpointRouting` option determines if routing should internally use endpoint-based logic or the legacy <xref:Microsoft.AspNetCore.Routing.IRouter>-based logic of ASP.NET Core 2.1 or earlier. When the compatibility version is set to 2.2 or later, the default value is `true`. Set the value to `false` to use the legacy routing logic:
+
+```csharp
+services.AddMvc()
+    .SetCompatibilityVersion(CompatibilityVersion.Version_2_2)
+    // Use the legacy routing of ASP.NET Core 2.1 or earlier:
+    .EnableEndpointRouting = false;
+```
```
Comment (rynowak) [inline aspnetcore/fundamentals/routing.md]: This isn't correct syntactically. It would look like:
```csharp
services.AddMvc(options => options.EnableEndpointRouting = false)
    .SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
```

File: aspnetcore/fundamentals/routing.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rynowak) [reply aspnetcore/fundamentals/routing.md]: I'd suggest punching this up to make the guidance about conventional routes specific to browser-based web applications. We're not trying to recommend that web apps use attribute routing. This section is really about helping people keep it simple.

**Comment by rynowak** [comment]
> I'll need an engineering sign-off in order to proceed. No giant rush tho ... AFAICT, we're look'in real good on 2.2 work.

I'm looking now 👍 - sorry for the delay I was on vacation

File: aspnetcore/fundamentals/routing.md [hunk 10]
```diff
@@ -1,77 +1,326 @@
...
+The routing system has the following characteristics:
+
+* Route template syntax is used to define routes with tokenized route parameters.
+* Conventional-style and attribute-style endpoint configuration is permitted.
+* `IRouteConstraint` is used to determine whether a URL parameter contains a valid value for a given endpoint constraint.
+* Frameworks, such as MVC/Razor Pages, register all of their endpoints, which have a predictable implementation of routing scenarios.
+* The routing implementation makes routing decisions wherever desired in the middleware pipeline.
+* Middleware that appears after a Routing Middleware can inspect the result of the Routing Middleware's endpoint decision for a given request URI.
+* It's possible to enumerate all of the endpoints in the app anywhere in the middleware pipeline.
+* A response can use routing to generate URLs (for example, for redirection or links) based on endpoint information and thus avoid hard-coded URLs, which helps maintainability.
```
Comment (rynowak) [inline aspnetcore/fundamentals/routing.md]: ` A response can use` suggestion -> `An application can use`

File: aspnetcore/fundamentals/routing.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: I think I need a suggested revision ... not sure what part(s) you want to change exactly.

Remove? ...

> Attribute routing models the app's URL-space as a set of resources with operations represented by HTTP verbs. This means that many operations (for example, GET, POST) on the same logical resource use the same URL.

Remove? ...

> using \[attribute routing](xref:mvc/controllers/routing#attribute-routing)

... there isn't much more here about attribute routing.

File: aspnetcore/fundamentals/routing.md [hunk 11]
```diff
@@ -1,77 +1,326 @@
...
+| ------------------ | ------------------------------------------------------------------- |
+| `GetPathByAddress` | Generates a URI with an absolute path based on the provided values. |
+| `GetUriByAddress`  | Generates an absolute URI based on the provided values.             |
+
+> [!WARNING]
+> Pay attention to the following security implications of calling `LinkGenerator` methods:
+>
+> Use `GetUri*` extension methods with caution in an app configuration that doesn't validate the `Host` header of incoming requests. If the `Host` header of incoming requests isn't validated, untrusted request input can be sent back to the client in URIs in a view/page. We recommend that all production apps configure their server to validate the `Host` header against known valid values.
+>
+> Use `LinkGenerator` with caution in middleware in combination with `Map` or `MapWhen`. `Map*` changes the base path of the executing request, which affects the output of link generation. All of the `LinkGenerator` APIs allow specifying a base path. Always specify an empty base path to undo `Map*`'s affect on link generation.
```
Comment (rynowak) [inline aspnetcore/fundamentals/routing.md]: This isn't a security thing, it's a gotcha that's pretty easy to hit.

File: aspnetcore/fundamentals/routing.md [hunk 12]
```diff
@@ -1,77 +1,326 @@
...
+  });
+  ```
+
+  Suppose you generate a link to an action using the following route:
+
+  ```csharp
+  var link = Url.Action("ReadPost", "blog", new { id = 17, });
+  ```
+
+  With `IRouter`-based routing, this code generates a URI of `/blog/ReadPost/17`, which respects the casing of the provided route value. Endpoint routing in ASP.NET Core 2.2 or later produces `/Blog/ReadPost/17` ("Blog" is capitalized). Endpoint routing provides a parameter transformer interface called `IOutboundParameterTransformer` that can be used to customize this behavior globally or to apply different conventions for mapping URLs.
```
Comment (rynowak) [inline aspnetcore/fundamentals/routing.md]: `provides a parameter transformer interface` -> `provides a feature called`

File: aspnetcore/fundamentals/routing.md [hunk 13]
```diff
@@ -1,77 +1,326 @@
...
+| `GetUriByAddress`  | Generates an absolute URI based on the provided values.             |
+
+> [!WARNING]
+> Pay attention to the following security implications of calling `LinkGenerator` methods:
+>
+> Use `GetUri*` extension methods with caution in an app configuration that doesn't validate the `Host` header of incoming requests. If the `Host` header of incoming requests isn't validated, untrusted request input can be sent back to the client in URIs in a view/page. We recommend that all production apps configure their server to validate the `Host` header against known valid values.
+>
+> Use `LinkGenerator` with caution in middleware in combination with `Map` or `MapWhen`. `Map*` changes the base path of the executing request, which affects the output of link generation. All of the `LinkGenerator` APIs allow specifying a base path. Always specify an empty base path to undo `Map*`'s affect on link generation.
+
+A few differences exist between endpoint routing in ASP.NET Core 2.2 or later and earlier versions of routing in ASP.NET Core:
```
Comment (rynowak) [inline aspnetcore/fundamentals/routing.md]: Should this be in a separate section that describes differences between the two frameworks? I wouldn't expect most users to need to know this.

File: aspnetcore/fundamentals/routing.md (hunk 12 above)
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: I can't use the word "feature" .... it's been **_banned_** from topics. :skull: lol

I'll go with ...

> Endpoint routing provides the `IOutboundParameterTransformer` interface that ...

Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: *Yikes!* ... I'm glad you said something. I found ~4 more "features" instances in the topic. They don't just **_release you_** for using banned words ... it's *death by firing squad!* ... :boy: :gun: :smile: lol

File: aspnetcore/fundamentals/routing.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rynowak) [reply aspnetcore/fundamentals/routing.md]: I worked with @Rick-Anderson recently on a section of one of these docs that tries to provide simple and actionable guidance based on whether you're create a web app or an API. I don't really want to see that blown up.

[the remark this answers is not in the document]
Comment (Rick-Anderson) [reply aspnetcore/fundamentals/routing.md]: >I worked with @Rick-Anderson recently on a section of one of these docs that tries to provide simple and actionable guidance based on whether you're create a web app or an API. 

I think [this](https://github.com/aspnet/Docs/pull/9240/files) is the changes you're talking about.

@rynowak 	@JamesNK 	I don't know what **punching this up** and **I don't really want to see that blown up.** means. Does that mean you don't want to amplify that guidance? That's what's there is fine?  If not, can one of you suggest a new outline and specific changes.

[the remark this answers is not in the document]
Comment (rynowak) [reply aspnetcore/fundamentals/routing.md]: I'm saying that I thought we did a good job.

File: aspnetcore/fundamentals/routing.md [hunk 14]
```diff
@@ -1,77 +1,327 @@
...
 
-It's common to add additional terse routes to high traffic areas of the app in specialized situations (for example, blog, ecommerce) using [attribute routing](xref:mvc/controllers/routing#attribute-routing) or dedicated conventional routes.
+::: moniker-end
 
-Web APIs should use attribute routing to model the app's functionality as a set of resources where operations are represented by HTTP verbs. This means that many operations (for example, GET, POST) on the same logical resource will use the same URL. Attribute routing provides a level of control that is needed to carefully design an API's URL-space.
+::: moniker range="= aspnetcore-2.2"
+
+To use the latest routing scenarios in ASP.NET Core 2.2, specify the [compatibility version](xref:mvc/compatibility-version) to the MVC services registration in `Startup.ConfigureServices`:
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: I didn't do that for this because the compat version (and the "legacy" option) is only applicable to 2.2. When 3.0 rolls out, these changes will be built-in, and no compat switch will be required to get endpoint routing.

WRT to "legacy" routing (and I'll react to your suggestion below in a sec) ... same thing. My suggestion is that this topic should jettison "legacy" routing language at 3.0. There won't be a valid reason to use `IRouter`-based routing at 3.0 AFAICT ... unless they say otherwise. btw ---- that word choice ---- "legacy" ---- that was Ryan's description of it ... I just used his language from his notes. I guess that if engineering views it as "legacy" tech at 2.2, they'll want to express it that way to the community. It does give one the feel'in that they should move away from it asap. Let me know if you would like it changed in spite of these notions ... I'm really fine either way on it.

File: aspnetcore/fundamentals/routing.md [hunk 15]
```diff
@@ -1,77 +1,327 @@
...
+
+```csharp
+services.AddMvc()
+    .SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
+```
+
+The `EnableEndpointRouting` option determines if routing should internally use endpoint-based logic or the legacy <xref:Microsoft.AspNetCore.Routing.IRouter>-based logic of ASP.NET Core 2.1 or earlier. When the compatibility version is set to 2.2 or later, the default value is `true`. Set the value to `false` to use the legacy routing logic:
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: :point_up: ... yeah ... Ryan's word for it. I assume that they would like folks off of it, certainly by 3.0. I opened this to address it :point_right: https://github.com/aspnet/Docs/issues/9552

File: aspnetcore/fundamentals/routing.md [hunk 16]
```diff
@@ -1,77 +1,327 @@
...
+> This document covers low-level ASP.NET Core routing. For information on ASP.NET Core MVC routing, see <xref:mvc/controllers/routing>. For information on routing conventions in Razor Pages, see <xref:razor-pages/razor-pages-conventions>.
 
 [View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/fundamentals/routing/samples) ([how to download](xref:index#how-to-download-a-sample))
 
 ## Routing basics
 
+Most apps should choose a basic and descriptive routing scheme so that URLs are readable and meaningful. The default conventional route `{controller=Home}/{action=Index}/{id?}`:
+
+* Supports a basic and descriptive routing scheme.
+* Is a useful starting point for apps that interact with browsers.
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: @rynowak @Rick-Anderson Given that we just had a back-and-forth over this section of the topic, I'm concerned about changing it again. Let me know if you're cool with the change here.

File: aspnetcore/fundamentals/routing.md [hunk 17]
```diff
@@ -1,77 +1,327 @@
...
+
+Developers commonly add additional terse routes to high traffic areas of an app in specialized situations (for example, blog and ecommerce endpoints) using [attribute routing](xref:mvc/controllers/routing#attribute-routing) or dedicated conventional routes.
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: On the dash, :+1:

On the rest, I yield to @rynowak and @Rick-Anderson for approval given their satisfaction with what was there.

File: aspnetcore/fundamentals/routing.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (scottaddie) [reply aspnetcore/fundamentals/routing.md]: My concern is around using the word "legacy" to describe a feature in an LTS release.

File: aspnetcore/fundamentals/routing.md [hunk 18]
```diff
@@ -1,77 +1,327 @@
...
+
+Web APIs should use attribute routing to model the app's functionality as a set of resources where operations are represented by HTTP verbs. This means that many operations (for example, GET, POST) on the same logical resource will use the same URL. Attribute routing provides a level of control that is needed to carefully design an API's URL-space.
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: @rynowak @Rick-Anderson ... again, your words here.

File: aspnetcore/fundamentals/routing.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/routing.md]: Good point. I'd prefer not to get called as a witness, too! :smile: lol

**Comment by guardrex** [comment]
Thanks @scottaddie ... I just need to hear back on that one section under routing basics. I ended up trying to change the existing text as little as possible. They like that section in the current version of the topic, so I reverted several of my updates in favor of what they had there. I think I licked everything else u listed.

**Comment by guardrex** [comment]
@scottaddie I'll give it one more pass in the morning ... see if I spot any sneaky 🐞 in here.

**Comment by guardrex** [comment]
I should take a final look at the "URI" vs "URL" situation. I thought there were at least a couple of spots where URI worked better and would be more consistent with notes and prior text. I'll ping u tomorrow @scottaddie with the commit (if there is one), and we can take a look at anything I find that might need to change.

**Comment by guardrex** [comment]
@rynowak @JamesNK Pain point here groking "URI" (25 instances) vs. "URL" (95 instances) in all contexts down the doc. For example, in the *URL generation* section, it starts with 'generating a URL' language ...

> routing can create a URL path

> separation between your endpoints and the URLs that access them

... flips over to URI ...

> they generate a URI containing

> generate a URI in the context of

... with extension methods named with "URI" ... `GetUriByAddress`.

Next section ...

> this code generates a URI of `/blog/ReadPost/17`

... but that's a relative URL only valid in the context of an actual URL with a protocol.

Later in the topic, we're back to "URL" generation lingo ...

> URL generation is the process by which routing can create a URL path based on a set of route values.

btw -- Should `Url.Action` and `Url.Page` be `IUrlHelper.Action` and `IUrlHelper.Page`?

## The TL;DR Question :smile:

What approach should be taken here to assess "URL" and "URI" in the topic, if anything? If you reviewed with this in mind and are ok with what you saw, then we can go ahead and merge this. I'm a bit concerned about it tho ... interchangeable back-and-forth use of them anyway, if that's what's happening in spots.

**Comment by Rick-Anderson** [comment]
@guardrex 	do what's need to get @scottaddie 	to approve then merge. You could create a P3 issue on URI vs URL, but I don't think that's necessary.

**Comment by guardrex** [comment]
I like it ... let's do that. It's obviously not a 2.2 priority, even if there is work to do on that aspect.

@scottaddie I think your feedback has been addressed. Anything else to do?
